File: src/algorithm/leetcode/Problem287.java
package algorithm.leetcode;

public class Problem287 {

    public static int findDuplicate(int[] nums) {
        int slow = 0, fast = 0, t = 0;
        while (true) {
            slow = nums[slow];
            fast = nums[nums[fast]];
            if (slow == fast)
                break;
        }
        while (true) {
            slow = nums[slow];
            t = nums[t];
            if (slow == t) break;
        }
        return slow;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{1,3,4,2,2};
        System.out.println(findDuplicate(nums));
    }

}

File: src/algorithm_cpp/problem287.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
	unordered_map<int, int> edges;
	for (int i = 0; i < nums.size(); i++) {
	    edges[i] = nums[i];
	}
	int fast = 0, slow = 0;
	while (fast != slow) {
	    fast = edges[fast];
	    fast = edges[fast];
	    slow = edges[slow];
	}
	int ptr = 0;
	while (ptr != slow) {
	    slow = edges[slow];
	    ptr = edges[ptr];
	}
	return ptr;
    }
};
